schemaless/sources.py

```diff
@@ -510,11 +527,62 @@ class AffordableRentalPortfolio(Source):
     DATA_SF = "https://data.sfgov.org/Housing-and-Buildings/Mayor-s-Office-of-Housing-and-Community-Developmen/9rdx-httc"  # NOQA
 
 
+class PermitAddendaSummary(Source):
+    '''Class for permit addenda summary data.
+
+    Note that the fields of this source do *not* necessarily map directly to
+    fields from the input file (unlike subclasses of DirectSource).
+
+    Instead, we look at all the addenda pertaining to a given permit, summarize
+    the  data into a few key fields, and output one record per permit with
+    those key fields.
+    '''
+    NAME = 'permit_addenda_summary'
+    FK = PrimaryKey(NAME, 'permit_number')
+
+    FIELDS = [
+        'permit_number',
+        'earliest_addenda_arrival'
+    ]
+
+    def field_names(self):
+        return self.FIELDS
+
+    def yield_records(self):
+        '''Outputs one record per permit number with a few key fields summarizing
+        info about related addenda.'''
+        with open_file(self._filepath,
+                       mode='rt',
+                       encoding='utf-8',
+                       errors='replace') as inf:
+            reader = DictReader(inf)
+
+            permit_to_arrival = {}
+            for line in reader:
+                permit_number = line['APPLICATION_NUMBER']
+                try:
+                    arrive_date = datetime.strptime(
+                        line['ARRIVE'].split(" ")[0], '%Y/%m/%d').date()
+                except ValueError:
+                    arrive_date = date.max
+                if permit_number not in permit_to_arrival or \
+                   arrive_date < permit_to_arrival[permit_number]:
+                    permit_to_arrival[permit_number] = arrive_date
+
+            for permit_number in permit_to_arrival.keys():
+                arrive_date = permit_to_arrival[permit_number]
+                arrive_date_str = arrive_date.isoformat() \
+                    if arrive_date != date.max else ''
+                values = [permit_number, arrive_date_str]
+                yield dict(zip(self.FIELDS, values))
```

Comment:
while this is pythonic, having something that depends on the order of self.FIELDS so may lines away from its declaration is asking for bugs, IMO. I think you should just build the dict manually.